### ISSUE TITLE
Upgrade gems to avoid vulnerabilities

### ISSUE THREAD

**Issue by matschaffer** [Safecast/safecastapi#191]
Came across https://isitvulnerable.com/results/b53abed48b0c8ae05497f8585d8ae61d to help with this

**Comment by matschaffer** [comment]
Here's the latest, though I'm tempted to close this as it'll likely get covered by the rails upgrades as we progress.

```json
{
  "vulnerable": true,
  "data": [
    {
      "type": "artifact-version",
      "attributes": {
        "name": "uglifier",
        "kind": "rubygem",
        "number": "2.0.1",
        "arch": null,
        "upgrade-to": ">= 2.7.2",
        "vulnerabilities": [
          {
            "description": "The upstream library for the Ruby uglifier gem, UglifyJS, is\naffected by a vulnerability that allows a specially crafted \nJavascript file to have altered functionality after minification.\n\nThis bug, found in UglifyJS versions 2.4.23 and earlier, was demonstrated\nto allow potentially malicious code to be hidden within secure code, \nand activated by the minification process.\n\nFor more information, consult: https://zyan.scripts.mit.edu/blog/backdooring-js/",
            "osvdb": [
              "126747"
            ],
            "upgrade-to": [
              ">= 2.7.2"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [],
            "title": "uglifier incorrectly handles non-boolean comparisons during minification",
            "uuid": "55de50a0-8e13-4ec0-aa1d-2715846c6e55",
            "unaffected-versions": [],
            "patched-versions": [
              ">= 2.7.2"
            ]
          }
        ]
      }
    },
    {
      "type": "artifact-version",
      "attributes": {
        "name": "actionpack",
        "kind": "rubygem",
        "number": "4.0.13",
        "arch": null,
        "upgrade-to": "~> 4.1.14, >= 4.1.14.2",
        "vulnerabilities": [
          {
            "description": "There is a possible object leak which can lead to a denial of service \nvulnerability in Action Pack. This vulnerability has been \nassigned the CVE identifier CVE-2016-0751. \n\nVersions Affected:  All. \nNot affected:       None. \nFixed Versions:     5.0.0.beta1.1, 4.2.5.1, 4.1.14.1, 3.2.22.1 \n\nImpact \n------ \nA carefully crafted accept header can cause a global cache of mime types to \ngrow indefinitely which can lead to a possible denial of service attack in \nAction Pack. \n\nAll users running an affected release should either upgrade or use one of the \nworkarounds immediately. \n\nReleases \n-------- \nThe FIXED releases are available at the normal locations. \n\nWorkarounds \n----------- \nThis attack can be mitigated by a proxy that only allows known mime types in \nthe Accept header. \n\nPlacing the following code in an initializer will also mitigate the issue: \n\n```ruby \nrequire 'action_dispatch/http/mime_type' \n\nMime.const_set :LOOKUP, Hash.new { |h,k| \n  Mime::Type.new(k) unless k.blank? \n} \n``` \n\nPatches \n------- \nTo aid users who aren't able to upgrade immediately we have provided patches for \nthe two supported release series. They are in git-am format and consist of a \nsingle changeset. \n\n* 5-0-mime_types_leak.patch - Patch for 5.0 series \n* 4-2-mime_types_leak.patch - Patch for 4.2 series \n* 4-1-mime_types_leak.patch - Patch for 4.1 series \n* 3-2-mime_types_leak.patch - Patch for 3.2 series \n\nPlease note that only the 4.1.x and 4.2.x series are supported at present. Users \nof earlier unsupported releases are advised to upgrade as soon as possible as we \ncannot guarantee the continued availability of security fixes for unsupported \nreleases. \n\nCredits \n------- \nAaron Patterson <3<3",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 4.2.5, >= 4.2.5.1",
              "~> 5.0.0.beta1.1"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2016-0751"
            ],
            "title": "Possible Object Leak and Denial of Service attack in Action Pack",
            "uuid": "56a6d4bf-522b-40c3-bffd-cc3c42304455",
            "unaffected-versions": [],
            "patched-versions": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 3.2.22.1",
              "~> 5.0.0.beta1.1",
              "~> 4.2.5, >= 4.2.5.1"
            ]
          },
          {
            "description": "There is a possible remote code execution vulnerability in Action Pack.\nThis vulnerability has been assigned the CVE identifier CVE-2016-2098.\n\nVersions Affected:  3.2.x, 4.0.x, 4.1.x, 4.2.x\nNot affected:       5.0+\nFixed Versions:     3.2.22.2, 4.1.14.2, 4.2.5.2\n\nImpact \n------ \nApplications that pass unverified user input to the `render` method in a\ncontroller or a view may be vulnerable to a code injection.\n\nImpacted code will look like this:\n\n```ruby\nclass TestController < ApplicationController\n  def show\n    render params[:id]\n  end\nend\n```\n\nAn attacker could use the request parameters to coerce the above example\nto execute arbitrary ruby code.\n\nAll users running an affected release should either upgrade or use one of \nthe workarounds immediately.\n\nReleases \n-------- \nThe FIXED releases are available at the normal locations.\n\nWorkarounds \n----------- \nA workaround to this issue is to not pass arbitrary user input to the `render`\nmethod. Instead, verify that data before passing it to the `render` method.\n\nFor example, change this:\n\n```ruby\ndef index\n  render params[:id]\nend\n```\n\nTo this:\n\n```ruby\ndef index\n  render verify_template(params[:id])\nend\n\nprivate\ndef verify_template(name)\n  # add verification logic particular to your application here\nend\n```\n\nPatches \n------- \nTo aid users who aren't able to upgrade immediately we have provided a \npatch for it. It is in git-am format and consist of a single changeset.\n\n* 3-2-secure_inline_with_params.patch - Patch for 3.2 series\n* 4-1-secure_inline_with_params.patch - Patch for 4.1 series\n* 4-2-secure_inline_with_params.patch - Patch for 4.2 series\n\nCredits \n------- \nThanks to both Tobias Kraze from makandra and joernchen of Phenoelit for \nreporting this!",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.14, >= 4.1.14.2",
              "~> 4.2.5, >= 4.2.5.2"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2016-2098"
            ],
            "title": "Possible remote code execution vulnerability in Action Pack",
            "uuid": "56d510a7-ad8a-4a61-a39e-904f5c6e2e05",
            "unaffected-versions": [
              ">= 5.0.0.beta1"
            ],
            "patched-versions": [
              "~> 4.2.5, >= 4.2.5.2",
              "~> 4.1.14, >= 4.1.14.2",
              "~> 3.2.22.2"
            ]
          },
          {
            "description": "There is a timing attack vulnerability in the basic authentication support \nin Action Controller. This vulnerability has been assigned the CVE \nidentifier CVE-2015-7576. \n\nVersions Affected:  All. \nNot affected:       None. \nFixed Versions:     5.0.0.beta1.1, 4.2.5.1, 4.1.14.1, 3.2.22.1 \n\nImpact \n------ \nDue to the way that Action Controller compares user names and passwords in \nbasic authentication authorization code, it is possible for an attacker to \nanalyze the time taken by a response and intuit the password. \n\nFor example, this string comparison: \n\n  \"foo\" == \"bar\" \n\nis possibly faster than this comparison: \n\n  \"foo\" == \"fo1\" \n\nAttackers can use this information to attempt to guess the username and \npassword used in the basic authentication system. \n\nYou can tell you application is vulnerable to this attack by looking for \n`http_basic_authenticate_with` method calls in your application. \n\nAll users running an affected release should either upgrade or use one of \nthe workarounds immediately. \n\nReleases \n-------- \nThe FIXED releases are available at the normal locations. \n\nWorkarounds \n----------- \nIf you can't upgrade, please use the following monkey patch in an initializer \nthat is loaded before your application: \n\n``` \n$ cat config/initializers/basic_auth_fix.rb \nmodule ActiveSupport \n  module SecurityUtils \n    def secure_compare(a, b) \n      return false unless a.bytesize == b.bytesize \n\n      l = a.unpack \"C#{a.bytesize}\" \n\n      res = 0 \n      b.each_byte { |byte| res |= byte ^ l.shift } \n      res == 0 \n    end \n    module_function :secure_compare \n\n    def variable_size_secure_compare(a, b) \n      secure_compare(::Digest::SHA256.hexdigest(a), ::Digest::SHA256.hexdigest(b)) \n    end \n    module_function :variable_size_secure_compare \n  end \nend \n\nmodule ActionController \n  class Base \n    def self.http_basic_authenticate_with(options = {}) \n      before_action(options.except(:name, :password, :realm)) do \n        authenticate_or_request_with_http_basic(options[:realm] || \"Application\") do |name, password| \n          # This comparison uses & so that it doesn't short circuit and \n          # uses `variable_size_secure_compare` so that length information \n          # isn't leaked. \n          ActiveSupport::SecurityUtils.variable_size_secure_compare(name, options[:name]) & \n            ActiveSupport::SecurityUtils.variable_size_secure_compare(password, options[:password]) \n        end \n      end \n    end \n  end \nend \n``` \n\n\nPatches \n------- \nTo aid users who aren't able to upgrade immediately we have provided patches for \nthe two supported release series. They are in git-am format and consist of a \nsingle changeset. \n\n* 4-1-basic_auth.patch - Patch for 4.1 series \n* 4-2-basic_auth.patch - Patch for 4.2 series \n* 5-0-basic_auth.patch - Patch for 5.0 series \n\nPlease note that only the 4.1.x and 4.2.x series are supported at present. Users \nof earlier unsupported releases are advised to upgrade as soon as possible as we \ncannot guarantee the continued availability of security fixes for unsupported \nreleases. \n\nCredits \n------- \n\nThank you to Daniel Waterworth for reporting the problem and working with us to \nfix it.",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 4.2.5, >= 4.2.5.1",
              "~> 5.0.0.beta1.1"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2015-7576"
            ],
            "title": "Timing attack vulnerability in basic authentication in Action Controller.",
            "uuid": "56a6d4bf-f6e1-4622-9fac-ba1d1bd6d254",
            "unaffected-versions": [],
            "patched-versions": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 3.2.22.1",
              "~> 5.0.0.beta1.1",
              "~> 4.2.5, >= 4.2.5.1"
            ]
          },
          {
            "description": "There is an object leak vulnerability for wildcard controllers in Action Pack. \nThis vulnerability has been assigned the CVE identifier CVE-2015-7581. \n\nVersions Affected:  >= 4.0.0 and < 5.0.0.beta1 \nNot affected:       < 4.0.0, 5.0.0.beta1 and newer \nFixed Versions:     4.2.5.1, 4.1.14.1 \n\nImpact \n------ \nUsers that have a route that contains the string \":controller\" are susceptible \nto objects being leaked globally which can lead to unbounded memory growth. \nTo identify if your application is vulnerable, look for routes that contain \n\":controller\". \n\nInternally, Action Pack keeps a map of \"url controller name\" to \"controller \nclass name\".  This map is cached globally, and is populated even if the \ncontroller class doesn't actually exist. \n\nAll users running an affected release should either upgrade or use one of the \nworkarounds immediately. \n\nReleases \n-------- \nThe FIXED releases are available at the normal locations. \n\nWorkarounds \n----------- \nThere are no feasible workarounds for this issue. \n\nPatches \n------- \nTo aid users who aren't able to upgrade immediately we have provided patches for the two supported release series.  They are in git-am format and consist of a single changeset. \n\n* 4-1-wildcard_route.patch - Patch for 4.1 series \n* 4-2-wildcard_route.patch - Patch for 4.2 series \n\nPlease note that only the 4.1.x and 4.2.x series are supported at present.  Users of earlier unsupported releases are advised to upgrade as soon as possible as we cannot guarantee the continued availability of security fixes for unsupported releases.",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 4.2.5, >= 4.2.5.1"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2015-7581"
            ],
            "title": "Object leak vulnerability for wildcard controller routes in Action Pack",
            "uuid": "56a6d4bf-7a66-4c29-9904-91b8c0a7dc0f",
            "unaffected-versions": [
              ">= 5.0.0.beta1",
              "< 4.0.0"
            ],
            "patched-versions": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 4.2.5, >= 4.2.5.1"
            ]
          }
        ]
      }
    },
    {
      "type": "artifact-version",
      "attributes": {
        "name": "activerecord",
        "kind": "rubygem",
        "number": "4.0.13",
        "arch": null,
        "upgrade-to": "~> 4.1.14, >= 4.1.14.1",
        "vulnerabilities": [
          {
            "description": "There is a vulnerability in how the nested attributes feature in Active Record \nhandles updates in combination with destroy flags when destroying records is \ndisabled. This vulnerability has been assigned the CVE identifier CVE-2015-7577. \n\nVersions Affected:  3.1.0 and newer \nNot affected:       3.0.x and older \nFixed Versions:     5.0.0.beta1.1, 4.2.5.1, 4.1.14.1, 3.2.22.1 \n\nImpact \n------ \nWhen using the nested attributes feature in Active Record you can prevent the \ndestruction of associated records by passing the `allow_destroy: false` option \nto the `accepts_nested_attributes_for` method. However due to a change in the \ncommit [a9b4b5d][1] the `_destroy` flag prevents the `:reject_if` proc from \nbeing called because it assumes that the record will be destroyed anyway. \n\nHowever this isn't true if `:allow_destroy` is false so this leads to changes \nthat would have been rejected being applied to the record. Attackers could use \nthis do things like set attributes to invalid values and to clear all of the \nattributes amongst other things. The severity will be dependent on how the \napplication has used this feature. \n\nAll users running an affected release should either upgrade or use one of \nthe workarounds immediately. \n\nReleases \n-------- \nThe FIXED releases are available at the normal locations. \n\nWorkarounds \n----------- \nIf you can't upgrade, please use the following monkey patch in an initializer \nthat is loaded before your application: \n\n``` \n$ cat config/initializers/nested_attributes_bypass_fix.rb \nmodule ActiveRecord \n  module NestedAttributes \n    private \n\n    def reject_new_record?(association_name, attributes) \n      will_be_destroyed?(association_name, attributes) || call_reject_if(association_name, attributes) \n    end \n\n    def call_reject_if(association_name, attributes) \n      return false if will_be_destroyed?(association_name, attributes) \n\n      case callback = self.nested_attributes_options[association_name][:reject_if] \n      when Symbol \n        method(callback).arity == 0 ? send(callback) : send(callback, attributes) \n      when Proc \n        callback.call(attributes) \n      end \n    end \n\n    def will_be_destroyed?(association_name, attributes) \n      allow_destroy?(association_name) && has_destroy_flag?(attributes) \n    end \n\n    def allow_destroy?(association_name) \n      self.nested_attributes_options[association_name][:allow_destroy] \n    end \n  end \nend \n``` \n\nPatches \n------- \nTo aid users who aren't able to upgrade immediately we have provided patches for \nthe two supported release series. They are in git-am format and consist of a \nsingle changeset. \n\n* 3-2-nested-attributes-reject-if-bypass.patch - Patch for 3.2 series \n* 4-1-nested-attributes-reject-if-bypass.patch - Patch for 4.1 series \n* 4-2-nested-attributes-reject-if-bypass.patch - Patch for 4.2 series \n* 5-0-nested-attributes-reject-if-bypass.patch - Patch for 5.0 series \n\nPlease note that only the 4.1.x and 4.2.x series are supported at present. Users \nof earlier unsupported releases are advised to upgrade as soon as possible as we \ncannot guarantee the continued availability of security fixes for unsupported \nreleases. \n\nCredits \n------- \nThank you to Justin Coyne for reporting the problem and working with us to fix it. \n\n[1]: https://github.com/rails/rails/commit/a9b4b5da7c216e4464eeb9dbd0a39ea258d64325",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 4.2.5, >= 4.2.5.1",
              "~> 5.0.0.beta1.1"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2015-7577"
            ],
            "title": "Nested attributes rejection proc bypass in Active Record",
            "uuid": "56a6d4bf-a33b-47d9-90c0-715f9ba68b45",
            "unaffected-versions": [
              "~> 3.0.0",
              "< 3.0.0"
            ],
            "patched-versions": [
              "~> 4.1.14, >= 4.1.14.1",
              "~> 3.2.22.1",
              "~> 5.0.0.beta1.1",
              "~> 4.2.5, >= 4.2.5.1"
            ]
          }
        ]
      }
    },
    {
      "type": "artifact-version",
      "attributes": {
        "name": "redcarpet",
        "kind": "rubygem",
        "number": "2.2.2",
        "arch": null,
        "upgrade-to": ">= 3.2.3",
        "vulnerabilities": [
          {
            "description": "redcarpet Gem for Ruby contains a flaw that allows a cross-site scripting\n(XSS) attack. This flaw exists because the parse_inline() function in\nmarkdown.c does not validate input before returning it to users. This may\nallow a remote attacker to create a specially crafted request that would\nexecute arbitrary script code in a user's browser session within the trust\nrelationship between their browser and the server.",
            "osvdb": [
              "120415"
            ],
            "upgrade-to": [
              ">= 3.2.3"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [],
            "title": "redcarpet Gem for Ruby markdown.c parse_inline() Function XSS",
            "uuid": "557513c5-a667-4070-b464-af50813dbb77",
            "unaffected-versions": [],
            "patched-versions": [
              ">= 3.2.3"
            ]
          }
        ]
      }
    },
    {
      "type": "artifact-version",
      "attributes": {
        "name": "activesupport",
        "kind": "rubygem",
        "number": "4.0.13",
        "arch": null,
        "upgrade-to": "~> 4.1.11",
        "vulnerabilities": [
          {
            "description": "Specially crafted XML documents can cause applications to raise a \n`SystemStackError` and potentially cause a denial of service attack.  This \nonly impacts applications using REXML or JDOM as their XML processor.  Other \nXML processors that Rails supports are not impacted. \n\nAll users running an affected release should either upgrade or use one of the work arounds immediately.\n\nWorkarounds \n----------- \nUse an XML parser that is not impacted by this problem, such as Nokogiri or \nLibXML.  You can change the processor like this: \n\n  ActiveSupport::XmlMini.backend = 'Nokogiri' \n\nIf you cannot change XML parsers, then adjust \n`RUBY_THREAD_MACHINE_STACK_SIZE`.",
            "osvdb": [],
            "upgrade-to": [
              "~> 4.1.11",
              ">= 4.2.2"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2015-3227"
            ],
            "title": "Possible Denial of Service attack in Active Support",
            "uuid": "55807540-8a18-4f42-baf3-e0300eee90da",
            "unaffected-versions": [],
            "patched-versions": [
              "~> 3.2.22",
              ">= 4.2.2",
              "~> 4.1.11"
            ]
          }
        ]
      }
    },
    {
      "type": "artifact-version",
      "attributes": {
        "name": "nokogiri",
        "kind": "rubygem",
        "number": "1.5.11",
        "arch": null,
        "upgrade-to": "~> 1.6.6.4",
        "vulnerabilities": [
          {
            "description": "Nokogiri Gem for JRuby contains a flaw that is triggered when handling a root\nelement in an XML document. This may allow a remote attacker to cause a\nconsumption of memory resources.",
            "osvdb": [
              "118481"
            ],
            "upgrade-to": [
              ">= 1.6.3"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [],
            "title": "Nokogiri Gem for JRuby XML Document Root Element Handling Memory Consumption\nRemote DoS",
            "uuid": "55b7e9d7-2e75-49df-a174-1e11eef1dd42",
            "unaffected-versions": [],
            "patched-versions": [
              ">= 1.6.3"
            ]
          },
          {
            "description": "Several vulnerabilities were discovered in the libxml2 and libxslt libraries\nthat the Nokogiri gem depends on.\n\nCVE-2015-1819\nA denial of service flaw was found in the way libxml2 parsed XML\ndocuments. This flaw could cause an application that uses libxml2 to use an\nexcessive amount of memory.\n\nCVE-2015-7941\nlibxml2 does not properly stop parsing invalid input, which allows\ncontext-dependent attackers to cause a denial of service (out-of-bounds read\nand libxml2 crash) via crafted specially XML data.\n\nCVE-2015-7942\nThe xmlParseConditionalSections function in parser.c in libxml2\ndoes not properly skip intermediary entities when it stops parsing invalid\ninput, which allows context-dependent attackers to cause a denial of service\n(out-of-bounds read and crash) via crafted XML data.\n\nCVE-2015-7995\nThe xsltStylePreCompute function in preproc.c in libxslt 1.1.28 does not\ncheck whether the parent node is an element, which allows attackers to cause\na denial of service using a specially crafted XML document.\n\nCVE-2015-8035\nThe xz_decomp function in xzlib.c in libxml2 2.9.1 does not\nproperly detect compression errors, which allows context-dependent attackers\nto cause a denial of service (process hang) via crafted XML data.\n\nAnother vulnerability was discoverd in libxml2 that could cause parsing\nof unclosed comments to result in \"conditional jump or move depends on\nuninitialized value(s)\" and unsafe memory access. This issue does not have a\nCVE assigned yet. See related URLs for details. Patched in v1.6.7.rc4.",
            "osvdb": [],
            "upgrade-to": [
              "~> 1.6.6.4",
              ">= 1.6.7.rc4"
            ],
            "criticality": "unknown",
            "rhsa": [],
            "cve": [
              "CVE-2015-1819"
            ],
            "title": "Nokogiri gem contains several vulnerabilities in libxml2 and libxslt",
            "uuid": "565780b7-c415-4079-a9f2-d5872317e76c",
            "unaffected-versions": [],
            "patched-versions": [
              "~> 1.6.6.4",
              ">= 1.6.7.rc4"
            ]
          }
        ]
      }
    }
  ]
}

```

**Comment by matschaffer** [comment]
A number of these have been covered already. There will be more over time as shown on https://github.com/Safecast/safecastapi/network/dependencies.

The current list is mostly the rails version covered by https://github.com/Safecast/safecastapi/issues/417 so I'll close this one. Once Rails is upgraded bumping any others should be straightforward.
